core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelService.java

```diff
@@ -157,4 +182,27 @@ private Set<ClusterComponentView> getComponentsByRequiredProducts(Map<String, Cl
                 .filter(Objects::nonNull)
                 .collect(Collectors.toSet());
     }
+
+    @Retryable(value = Exception.class, maxAttemptsExpression = "${cb.parcel.retry.maxAttempts:5}",
+            backoff = @Backoff(delayExpression = "${cb.parcel.retry.backOffDelay:2000}",
+                    multiplierExpression = "${cb.parcel.retry.backOffMultiplier:2}"))
+    public Optional<Response> getHeadResponseForParcel(String url) {
+        Client client = restClientFactory.getOrCreateDefault();
+        WebTarget target = client.target(url);
+        paywallCredentialPopulator.populateWebTarget(url, target);
+        Response response = target.request().head();
+        LOGGER.info("Head request for {} status: {}", url, response.getStatus());
+        if (retryableHttpCodes.contains(response.getStatus())) {
+            if (RetrySynchronizationManager.getContext().getRetryCount() < retryMaxAttempts - 1) {
```

Comment:
maxAttemptsExpression will govern if retry needs to be kicked in based on exception thrown. For retryable Http codes,  I have to explicitly throw exception so that retry can be kicked in. This block has been added to return Http code instead of exception if retry has been exhausted. Let's say we keep getting retryable Http code. if we don't have this block, exception will be thrown instead of returning Response with Http code. I thought, we might need this Http code in future so it's better to return it instead of suppressing it via exception. If we are good to throw exception even in case of retryable Http codes , I can remove this block.

Reply:
Ok, that is ok for me